File: frontend/src/api/llm.ts
import request from './request';

export type Provider = 'openai' | 'azure' | 'qwen' | 'deepseek' | 'zhipu';
export type LlmSource = 'company' | 'personal';
export type AccessStatus = 'pending' | 'approved' | 'rejected';

/** 公司 LLM 配置（员工视角：只含掩码与启用状态） */
export interface CompanyLlmConfig {
  is_active: boolean;
  provider: Provider;
  base_url: string;
  api_key_masked: string;
  models: string[];
  monthly_budget: number;
}

/** 个人 LLM 配置 */
export interface PersonalLlmConfig {
  is_active: boolean;
  provider: Provider;
  base_url: string;
  api_key_masked: string;
  models: string[];
}

/** 保存配置请求体（apiKey 为明文，落库时加密） */
export interface SaveCompanyConfigPayload {
  is_active?: boolean;
  provider: Provider;
  base_url: string;
  api_key?: string; // 留空表示不修改
  models: string[];
  monthly_budget: number;
}

export interface SavePersonalConfigPayload {
  is_active?: boolean;
  provider: Provider;
  base_url: string;
  api_key: string;
  models: string[];
}

export interface AccessRequest {
  id: number;
  user_id: number;
  user_name: string;
  company_id: number;
  status: AccessStatus;
  reason?: string;
  created_at: string;
  reviewed_at?: string;
  reviewed_by?: number;
}

export interface UsageRecord {
  user_id: number;
  user_name: string;
  source: LlmSource;
  provider: Provider;
  model: string;
  calls: number;
  prompt_tokens: number;
  completion_tokens: number;
  cost: number;
  quota_limit: number;
}

/** 后端 /usage/me 实际返回的嵌套结构 */
export interface MyUsageResponse {
  records: UsageRecord[];
  totals: {
    calls: number;
    prompt_tokens: number;
    completion_tokens: number;
    cost: number;
  };
  quota: {
    period: string;
    quota_limit: number;
    used: number;
    status: string;
  } | null;
}

export const llmApi = {
  // 公司配置
  getCompanyConfig: () =>
    request.get<CompanyLlmConfig>('/llm/config/company').then((r) => r.data),
  saveCompanyConfig: (payload: SaveCompanyConfigPayload) =>
    request.put<CompanyLlmConfig>('/llm/config/company', payload).then((r) => r.data),

  // 个人配置
  getPersonalConfig: () =>
    request.get<PersonalLlmConfig>('/llm/config/me').then((r) => r.data),
  savePersonalConfig: (payload: SavePersonalConfigPayload) =>
    request.put<PersonalLlmConfig>('/llm/config/me', payload).then((r) => r.data),

  // 使用方式切换
  getSource: () =>
    request.get<{ source: LlmSource }>('/llm/source').then((r) => r.data),
  switchSource: (source: LlmSource) =>
    request.put<{ source: LlmSource }>('/llm/source', { source }).then((r) => r.data),

  // 申请使用公司 API
  createAccessRequest: (reason?: string) =>
    request.post<AccessRequest>('/llm/access-request', { reason }).then((r) => r.data),

  // 管理员：待审批列表
  listAccessRequests: () =>
    request.get<AccessRequest[]>('/llm/access-requests').then((r) => r.data),

  // 管理员：审批
  reviewAccessRequest: (id: number, approved: boolean) =>
    request
      .put<AccessRequest>(`/llm/access-requests/${id}`, { action: approved ? 'approve' : 'reject' })
      .then((r) => r.data),

  // 用量
  getMyUsage: () => request.get<MyUsageResponse>('/llm/usage/me').then((r) => r.data),
  getCompanyUsage: () =>
    request.get<UsageRecord[]>('/llm/usage/company').then((r) => r.data),

  // 管理员：设置额度
  setQuotaLimit: (userId: number, limit: number) =>
    request
      .put<{ ok: true }>(`/llm/quotas/${userId}`, { quota_limit: limit })
      .then((r) => r.data),
};

File: backend/app/schemas/llm.py
"""LLM 相关 Pydantic 模型。

约定：API Key 输入用明文（字段名 api_key），输出仅回掩码（字段名 api_key_masked，
格式 sk-••••f2a）。后端永不回传明文 Key。
"""
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field

Provider = Literal["openai", "azure", "qwen", "deepseek", "zhipu"]
LlmSource = Literal["company", "personal"]


# ---------------- 公司配置 ----------------
class CompanyLlmConfigIn(BaseModel):
    """管理员保存公司配置（API Key 明文入参）。"""

    is_active: bool = False
    provider: Provider = "openai"
    base_url: str = ""
    api_key: str = Field("", description="明文 API Key，落库前 AES 加密")
    models: list[str] = []
    monthly_budget: float = 0.0


class CompanyLlmConfigOut(BaseModel):
    """管理员视角：完整配置（Key 仍为掩码）。"""

    is_active: bool
    provider: Provider
    base_url: str
    api_key_masked: str
    models: list[str]
    monthly_budget: float


class CompanyLlmConfigPublic(BaseModel):
    """员工视角：仅启用状态与掩码（及 provider 上下文）。"""

    is_active: bool
    api_key_masked: str
    provider: Optional[Provider] = None


# ---------------- 个人配置 ----------------
class PersonalLlmConfigIn(BaseModel):
    is_active: bool = False
    provider: Provider = "deepseek"
    base_url: str = ""
    api_key: str = ""
    models: list[str] = []


class PersonalLlmConfigOut(BaseModel):
    is_active: bool
    provider: Provider
    base_url: str
    api_key_masked: str
    models: list[str]


# ---------------- 使用方式切换 ----------------
class LlmSourceOut(BaseModel):
    source: LlmSource


class LlmSourceIn(BaseModel):
    source: LlmSource


# ---------------- 公司 API 访问申请 ----------------
class AccessRequestCreate(BaseModel):
    reason: Optional[str] = None


class AccessRequestOut(BaseModel):
    id: int
    user_id: int
    user_name: str
    company_id: int
    status: Literal["pending", "approved", "rejected"]
    reason: Optional[str] = None
    created_at: datetime
    reviewed_by: Optional[int] = None
    reviewed_at: Optional[datetime] = None


class AccessRequestReview(BaseModel):
    action: Literal["approve", "reject"]


# ---------------- 用量统计 ----------------
class UsageRecordOut(BaseModel):
    user_id: int
    user_name: str
    source: LlmSource
    provider: Provider
    model: str
    calls: int
    prompt_tokens: int
    completion_tokens: int
    cost: float
    quota_limit: float


class MyUsageOut(BaseModel):
    records: list[UsageRecordOut]
    totals: dict
    quota: Optional[dict] = None


# ---------------- 额度 ----------------
class QuotaUpdate(BaseModel):
    quota_limit: float


class QuotaOut(BaseModel):
    user_id: int
    company_id: int
    period: str
    quota_limit: float
    used: float
    status: str
